algosdk/testing/teal_blackbox.py

```diff
@@ -0,0 +1,720 @@
+        for x in [t.get("scratch", []) for t in trace]
+    ]
+    scratches = [
+        {i: s for i, s in enumerate(scratch) if not s.is_empty()}
+        for scratch in scratches
+    ]
+    slots_used = sorted(set().union(*(s.keys() for s in scratches)))
+    final_scratch_state = scratches[-1]
+    if not scratch_verbose:
+
+        def compute_delta(prev, curr):
+            pks, cks = set(prev.keys()), set(curr.keys())
+            new_keys = cks - pks
+            if new_keys:
+                return {k: curr[k] for k in new_keys}
+            return {k: v for k, v in curr.items() if prev[k] != v}
+
+        scratch_deltas = [scratches[0]]
+        for i in range(1, len(scratches)):
+            scratch_deltas.append(
+                compute_delta(scratches[i - 1], scratches[i])
+            )
+
+        scratches = [
+            [f"{i}{scratch_colon}{v}" for i, v in scratch.items()]
+            for scratch in scratch_deltas
+        ]
+    else:
+        scratches = [
+            [
+                f"{i}{scratch_colon}{scratch[i]}" if i in scratch else ""
+                for i in slots_used
+            ]
+            for scratch in scratches
+        ]
+
+    assert N == len(
+        scratches
+    ), f"mismatch of lengths in tls v. scratches ({N} v. {len(scratches)})"
+
+    bbr = BlackBoxResults(
+        N,
+        pcs,
+        line_nums,
+        tls,
+        stacks,
+        scratches,
+        final_scratch_state,
+        slots_used,
+        raw_stacks,
+    )
+    bbr.assert_well_defined()
+    return bbr
+
+
+def get_blackbox_scenario_components(
+    scenario: Dict[str, Union[list, dict]], mode: ExecutionMode
+) -> Tuple[List[tuple], Dict[DRA, Any]]:
+    """
+    Validate that a Blackbox Test Scenario has been properly constructed, and return back
+    its components which consist of **inputs** and _optional_ **assertions**.
+
+    A scenario should adhere to the following schema:
+    ```
+    {
+        "inputs":       List[Tuple[Union[str, int]]],
+        "assertions":   Dict[DryRunAssertionType, ...an assertion...]
+    }
+
+    Each assertion is a map from _assertion type_  to be made on a dry run,
+    to the actual assertion. Actual assertions can be:
+    * simple python types - these are useful in the case of _constant_ assertions.
+        For example, if you want to assert that the `maxStackHeight` is 3, just use `3`.
+    * dictionaries of type Dict[Tuple, Any] - these are useful when you just want to assert
+        a discrete set of input-output pairs.
+        For example, if you have 4 inputs that you want to assert are being squared,
+        you could use `{(2,): 4, (7,): 49, (13,): 169, (11,): 121}`
+    * functions which take a single variable. These are useful when you have a python "simulator"
+        for the assertions.
+        In the square example you could use `lambda args: args[0]**2`
+    * functions which take _two_ variables. These are useuful when your assertion is more
```

Comment:
@tzaffi Friendly pointer to a minor typo.

```suggestion
    * functions which take _two_ variables. These are useful when your assertion is more
```